### PR TITLE
Update ports-of-call and introduce error handling

### PR DESCRIPTION
**Pull request by Yurlungur**
## PR Summary

This pulls my error handling machinery through `singularity-eos`, mostly motivated by @aematt's request for testing. I make a few changes of note:
1. I make `ports-of-call` its own submodule. This is unfortunately required for now. Once @mauneyc-LANL 's PR #221 is through, this won't be necessary and the submodule should be removed. Note that I **do not** update `spiner`. I think @mauneyc-LANL 's changes to `spiner` make it incompatible with the submodule path for now. So spiner stays on the old commit until #221 goes through.
2. I create a new macro, `REQUIRE_MAYBE_THROWS` which is to be used in units tests with catch. It either resolves to a `REQUIRE_THROWS` if `PORTABILITY_STRATEGY_NONE` or a no-op. This is what @aematts should use.
3. As a temp hack, I replace `EOS_ERROR` with `PORTABLE_ABORT`. We should move to `PORTABLE_*` everywhere with time. This is just for backwards compatibility.

## PR Checklist

- [x] Adds a test for any bugs fixed. Adds tests for new features.
- [x] Format your changes by using the `make format` command after configuring with `cmake`.
- [x] Document any new features, update documentation for changes made.
- [x] Make sure the copyright notice on any files you modified is up to date.
- [x] After creating a pull request, note it in the CHANGELOG.md file
- [x] If preparing for a new release, update the version in cmake.

### CODE REVIEW

**Comment by Yurlungur** [comment]
@aematts once this is merged in the way to get your tests to work is to use `PORTABLE_ALWAYS_THROW_OR_ABORT` in the places you want caught, then use `PORTABLE_REQUIRE_MAYBE_THROWS` in the test to catch it.

**Review by jhp-lanl — APPROVED** [review]
Looks pretty straight-forward. I approve

**Comment by Yurlungur** [comment]
Ok tests now pass on github and re-git. To make tests pass, I had to change the spack spec to point to the main branch of `ports-of-call`. This is temporary and can be removed when `ports-of-call` has a new updated spackage in spack upstream for the [newest release](https://github.com/lanl/ports-of-call/releases/tag/v1.4.0).

**Comment by Yurlungur** [comment]
@rbberger please take a look at the spack trickery I added here and let me know whether you're ok with it

**Review by aematts — APPROVED** [review]
Looks fine to me even though I am not sure how to use it yet. Once I start using it I might come with more comments.

File: spack-repo/packages/singularity-eos/package.py
Comment (rbberger) [inline spack-repo/packages/singularity-eos/package.py:66]: Should be
```suggestion
    depends_on("ports-of-call@1.4.0:")
```
You also want the same in `spiner`.

Comment (Yurlungur) [reply spack-repo/packages/singularity-eos/package.py:66]: Spiner isn't using the new code. I'll add the colon in singularity.

**Review by rbberger — APPROVED** [review]
Approving this, assuming the re-git CI passes.

File: test/test_eos_unit.cpp
Comment (dholladay00) [inline test/test_eos_unit.cpp:139]: Is catch ok when this scenario is empty as is the case if portability strategy none is not defined @Yurlungur ?

Comment (Yurlungur) [reply test/test_eos_unit.cpp:139]: Catch is ok with empty scenarios I think. The scenario just auto-passes with zero conditions I believe.

Comment (dholladay00) [reply test/test_eos_unit.cpp:139]: Is the empty scenario scenario tested in our CI @Yurlungur ?

Comment (Yurlungur) [reply test/test_eos_unit.cpp:139]: Yes if the CI passes on GPU it passed for the empty scenario.

**Comment by Yurlungur** [comment]
Tests pass on re-git. Pulling the trigger.

**Comment by aematts** [comment]
> @aematts once this is merged in the way to get your tests to work is to use `PORTABLE_ALWAYS_THROW_OR_ABORT` in the places you want caught, then use `REQUIRE_MAYBE_THROWS` in the test to catch it.

@Yurlungur , did you get something in for the necessary warning when T<0? You did not like to have a plain write statement so deep in the code.

**Comment by Yurlungur** [comment]
@aematts ah yes I did. You can use `PORTABLE_WARN`, which is a no-op by default, but when compiled in debug mode, it will turn into an actual warning.

**Pull request merged**
